### ISSUE TITLE
Deadlock on sleep

### ISSUE THREAD

**Issue by ktf** [FairRootGroup/FairMQ#137]
Whenever one `sleep`s in the ConditionalRun, FairMQ deadlocks with it at:

```C++
    frame #0: 0x00007fff73f4e7de libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fff74008593 libsystem_pthread.dylib`_pthread_cond_wait + 724
    frame #2: 0x00007fff714e7bda libc++.1.dylib`std::__1::condition_variable::wait(std::__1::unique_lock<std::__1::mutex>&) + 18
    frame #3: 0x0000000109f1b6cb libFairMQ.1.3.dylib`void fair::mq::fsm::Machine_::StopFct::operator(this=<unavailable>, (null)=<unavailable>, fsm=0x00007f9f8cd16f90, (null)=<unavailable>, ts=<unavailable
>)<fair::mq::fsm::STOP_FSM_EVENT, boost::msm::back::state_machine<fair::mq::fsm::Machine_, boost::parameter::void_, boost::parameter::void_, boost::parameter::void_, boost::parameter::void_>, fair::mq::fs
m::RUNNING_FSM_STATE, fair::mq::fsm::READY_FSM_STATE>(fair::mq::fsm::STOP_FSM_EVENT const&, boost::msm::back::state_machine<fair::mq::fsm::Machine_, boost::parameter::void_, boost::parameter::void_, boost
::parameter::void_, boost::parameter::void_>&, fair::mq::fsm::RUNNING_FSM_STATE&, fair::mq::fsm::READY_FSM_STATE&) at FairMQStateMachine.cxx:254 [opt]
    frame #4: 0x0000000109f1b63d libFairMQ.1.3.dylib`boost::msm::back::state_machine<fair::mq::fsm::Machine_, boost::parameter::void_, boost::parameter::void_, boost::parameter::void_, boost::parameter::v
oid_>::a_row_<boost::msm::front::Row<fair::mq::fsm::RUNNING_FSM_STATE, fair::mq::fsm::STOP_FSM_EVENT, fair::mq::fsm::READY_FSM_STATE, fair::mq::fsm::Machine_::StopFct, boost::msm::front::none> >::execute(
boost::msm::back::state_machine<fair::mq::fsm::Machine_, boost::parameter::void_, boost::parameter::void_, boost::parameter::void_, boost::parameter::void_>&, int, int, fair::mq::fsm::STOP_FSM_EVENT const
&) [inlined] boost::msm::back::HandledEnum boost::msm::front::Row<fair::mq::fsm::RUNNING_FSM_STATE, fair::mq::fsm::STOP_FSM_EVENT, fair::mq::fsm::READY_FSM_STATE, fair::mq::fsm::Machine_::StopFct, boost::
msm::front::none>::action_call<fair::mq::fsm::STOP_FSM_EVENT, boost::msm::back::state_machine<fair::mq::fsm::Machine_, boost::parameter::void_, boost::parameter::void_, boost::parameter::void_, boost::par
ameter::void_>, fair::mq::fsm::RUNNING_FSM_STATE, fair::mq::fsm::READY_FSM_STATE, boost::fusion::set<fair::mq::fsm::ERROR_FSM_STATE, fair::mq::fsm::EXITING_FSM_STATE, fair::mq::fsm::OK_FSM_STATE, fair::mq
::fsm::RESETTING_DEVICE_FSM_STATE, fair::mq::fsm::RESETTING_TASK_FSM_STATE, fair::mq::fsm::PAUSED_FSM_STATE, fair::mq::fsm::RUNNING_FSM_STATE, fair::mq::fsm::READY_FSM_STATE, fair::mq::fsm::INITIALIZING_T
ASK_FSM_STATE, fair::mq::fsm::DEVICE_READY_FSM_STATE, fair::mq::fsm::INITIALIZING_DEVICE_FSM_STATE, fair::mq::fsm::IDLE_FSM_STATE> >(fsm=0x00007f9f8cd16f90, evt=<unavailable>, src=<unavailable>, tgt=<unav
ailable>, (null)=0x00007f9f8cd17167) at functor_row.hpp:108 [opt]
```

This is the reason why some tests randomly timeout in O2. I think this is a regression that appeared when we moved to 1.3.7.

**Comment by dennisklein** [comment]
@ktf: Can you link to one of the `ConditionalRun`s that is affected by this issue?

**Comment by ktf** [comment]
Other thread:

```C++
    frame #0: 0x00007fff73f4eeae libsystem_kernel.dylib`__semwait_signal + 10
    frame #1: 0x00007fff73ed9830 libsystem_c.dylib`nanosleep + 199
    frame #2: 0x00007fff73ed9692 libsystem_c.dylib`sleep + 41
  * frame #3: 0x000000010d494aed test_Framework_test_SimpleStatefulProcessing01`defineDataProcessing(this=0x00007fbe73f186f8, outputs=0x00007fbe7602f440)::$_0::operator()(o2::framework::CallbackService&)
const::'lambda'(o2::framework::DataAllocator&)::operator()(o2::framework::DataAllocator&) const at test_SimpleStatefulProcessing01.cxx:55
    frame #4: 0x000000010d494abd test_Framework_test_SimpleStatefulProcessing01`void std::__1::__invoke_void_return_wrapper<void>::__call<defineDataProcessing(o2::framework::ConfigContext const&)::$_0::op
erator()(o2::framework::CallbackService&) const::'lambda'(o2::framework::DataAllocator&)&, o2::framework::DataAllocator&>(defineDataProcessing(o2::framework::ConfigContext const&)::$_0::operator()(o2::framework::CallbackService&) const::'lambda'(o2::framework::DataAllocator&)&&&, o2::framework::DataAllocator&&&) [inlined] decltype(__f=0x00007fbe73f186f8, __args=0x00007fbe7602f440)::$_0::operator()(o2::fra
mework::CallbackService&) const::'lambda'(o2::framework::DataAllocator&)&>(fp)(std::__1::forward<o2::framework::DataAllocator&>(fp0))) std::__1::__invoke<defineDataProcessing(o2::framework::ConfigContext
const&)::$_0::operator()(o2::framework::CallbackService&) const::'lambda'(o2::framework::DataAllocator&)&, o2::framework::DataAllocator&>(defineDataProcessing(o2::framework::ConfigContext const&)::$_0::op
erator()(o2::framework::CallbackService&) const::'lambda'(o2::framework::DataAllocator&)&&&, o2::framework::DataAllocator&&&) at type_traits:4428
    frame #5: 0x000000010d494aa0 test_Framework_test_SimpleStatefulProcessing01`void std::__1::__invoke_void_return_wrapper<void>::__call<defineDataProcessing(__args=0x00007fbe73f186f8, __args=0x00007fbe7
602f440)::$_0::operator()(o2::framework::CallbackService&) const::'lambda'(o2::framework::DataAllocator&)&, o2::framework::DataAllocator&>(defineDataProcessing(o2::framework::ConfigContext const&)::$_0::operator()(o2::framework::CallbackService&) const::'lambda'(o2::framework::DataAllocator&)&&&, o2::framework::DataAllocator&&&) at __functional_base:349
    frame #6: 0x000000010d494994 test_Framework_test_SimpleStatefulProcessing01`std::__1::__function::__func<defineDataProcessing(o2::framework::ConfigContext const&)::$_0::operator()(o2::framework::CallbackService&) const::'lambda'(o2::framework::DataAllocator&), std::__1::allocator<defineDataProcessing(o2::framework::ConfigContext const&)::$_0::operator()(o2::framework::CallbackService&) const::'lambda'
(o2::framework::DataAllocator&)>, void (o2::framework::DataAllocator&)>::operator(this=0x00007fbe73f186f0, __arg=0x00007fbe7602f440)(o2::framework::DataAllocator&) at functional:1562
    frame #7: 0x000000010d4938e9 test_Framework_test_SimpleStatefulProcessing01`std::__1::function<void (o2::framework::DataAllocator&)>::operator(this=0x00007fbe73f186f0, __arg=0x00007fbe7602f440)(o2::fr
amework::DataAllocator&) const at functional:1913
    frame #8: 0x000000010d493860 test_Framework_test_SimpleStatefulProcessing01`std::__1::function<void (o2::framework::ProcessingContext&)> o2::framework::adaptStatelessF<o2::framework::DataAllocator>(th
is=0x00007fbe73f186f0, ctx=0x00007ffee278f510)>)::'lambda'(o2::framework::ProcessingContext&)::operator()(o2::framework::ProcessingContext&) const at AlgorithmSpec.h:122
    frame #9: 0x000000010d49381d test_Framework_test_SimpleStatefulProcessing01`void std::__1::__invoke_void_return_wrapper<void>::__call<std::__1::function<void (o2::framework::ProcessingContext&)> o2::f
ramework::adaptStatelessF<o2::framework::DataAllocator>(std::__1::function<void (o2::framework::DataAllocator&)>)::'lambda'(o2::framework::ProcessingContext&)&, o2::framework::ProcessingContext&>(o2::fram
ework::DataAllocator&&) [inlined] decltype(__f=0x00007fbe73f186f0, __args=0x00007ffee278f510)> o2::framework::adaptStatelessF<o2::framework::DataAllocator>(std::__1::function<void (o2::framework::DataAllo
cator&)>)::'lambda'(o2::framework::ProcessingContext&)&>(fp)(std::__1::forward<o2::framework::ProcessingContext&>(fp0))) std::__1::__invoke<std::__1::function<void (o2::framework::ProcessingContext&)> o2:
:framework::adaptStatelessF<o2::framework::DataAllocator>(std::__1::function<void (o2::framework::DataAllocator&)>)::'lambda'(o2::framework::ProcessingContext&)&, o2::framework::ProcessingContext&>(std::_
_1::function<void (o2::framework::ProcessingContext&)> o2::framework::adaptStatelessF<o2::framework::DataAllocator>(std::__1::function<void (o2::framework::DataAllocator&)>)::'lambda'(o2::framework::Proce
ssingContext&)&&&, o2::framework::ProcessingContext&&&) at type_traits:4428
    frame #10: 0x000000010d493800 test_Framework_test_SimpleStatefulProcessing01`void std::__1::__invoke_void_return_wrapper<void>::__call<std::__1::function<void (__args=0x00007fbe73f186f0, __args=0x0000
7ffee278f510)> o2::framework::adaptStatelessF<o2::framework::DataAllocator>(std::__1::function<void (o2::framework::DataAllocator&)>)::'lambda'(o2::framework::ProcessingContext&)&, o2::framework::Processi
ngContext&>(o2::framework::DataAllocator&&) at __functional_base:349
    frame #11: 0x000000010d493509 test_Framework_test_SimpleStatefulProcessing01`std::__1::__function::__func<std::__1::function<void (o2::framework::ProcessingContext&)> o2::framework::adaptStatelessF<o2
::framework::DataAllocator>(std::__1::function<void (o2::framework::DataAllocator&)>)::'lambda'(o2::framework::ProcessingContext&), std::__1::allocator<std::__1::function<void (o2::framework::ProcessingCo
ntext&)> o2::framework::adaptStatelessF<o2::framework::DataAllocator>(std::__1::function<void (o2::framework::DataAllocator&)>)::'lambda'(o2::framework::ProcessingContext&)>, void (o2::framework::Processi
ngContext&)>::operator(this=0x00007fbe73f186e0, __arg=0x00007ffee278f510)(o2::framework::ProcessingContext&) at functional:1562
    frame #12: 0x000000010eb2a579 libFramework.dylib`std::__1::function<void (o2::framework::ProcessingContext&)>::operator(this=0x00007fbe73f186e0, __arg=0x00007ffee278f510)(o2::framework::ProcessingCont
ext&) const at functional:1913
    frame #13: 0x000000010eb1f797 libFramework.dylib`o2::framework::DataProcessingDevice::tryDispatchComputation(this=0x00007ffee2790508, slot=(index = 14), record=0x00007ffee2790fc0)::$_11::operator()(o2
::framework::TimesliceSlot, o2::framework::InputRecord&) const at DataProcessingDevice.cxx:344
    frame #14: 0x000000010eb1701e libFramework.dylib`o2::framework::DataProcessingDevice::tryDispatchComputation(this=0x00007fbe7602f000) at DataProcessingDevice.cxx:522
    frame #15: 0x000000010eb15bab libFramework.dylib`o2::framework::DataProcessingDevice::ConditionalRun(this=0x00007fbe7602f000) at DataProcessingDevice.cxx:151
    frame #16: 0x000000010f64d04e libFairMQ.1.3.dylib`FairMQDevice::RunWrapper(this=0x00007fbe7602f000) at FairMQDevice.cxx:375 [opt]
    frame #17: 0x000000010f65f6eb libFairMQ.1.3.dylib`fair::mq::fsm::Machine_::ProcessWork() [inlined] std::__1::function<void ()>::operator(this=<unavailable>)() const at functional:1913 [opt]
    frame #18: 0x000000010f65f6d5 libFairMQ.1.3.dylib`fair::mq::fsm::Machine_::ProcessWork(this=0x00007fbe73e200d0) at FairMQStateMachine.cxx:392 [opt]
    frame #19: 0x000000010f65f447 libFairMQ.1.3.dylib`FairMQStateMachine::ProcessWork(this=0x00007fbe7602f000) at FairMQStateMachine.cxx:656 [opt]
    frame #20: 0x000000010f629e48 libFairMQ.1.3.dylib`fair::mq::DeviceRunner::Run() [inlined] FairMQDevice::RunStateMachine(this=0x00007fbe7602f000) at FairMQDevice.h:393 [opt]
    frame #21: 0x000000010f629e33 libFairMQ.1.3.dylib`fair::mq::DeviceRunner::Run(this=0x00007ffee27939d8) at DeviceRunner.cxx:101 [opt]
    frame #22: 0x000000010ecc483a libFramework.dylib`doChild(argc=19, argv=0x00007ffee279d248, spec=0x00007fbe77011200) at runDataProcessing.cxx:672
    frame #23: 0x000000010ecca498 libFramework.dylib`runStateMachine(workflow=size=3, driverControl=0x00007ffee279b208, driverInfo=0x00007ffee279b090, metricsInfos=size=0, frameworkId="producer") at runDa
```

**Comment by ktf** [comment]
https://github.com/AliceO2Group/AliceO2/blob/dev/Framework/Core/src/DataProcessingDevice.cxx

**Comment by ktf** [comment]
As I discussed privately, I would still vote for a completely threadless, synchronous and cooperative, `FairMQDeviceBase` version which exposes a `Collaborate()` method for the user to give back control to the device, so that we are not forced to have control threads and alike in case code runs in any case under a watchdog. The actual `FairMQDevice` class could actually be implemented on top of it.

**Comment by rbx** [comment]
Looking at the second thread - it is blocked in the `sleep`, which prevents `ConditionalRun` from finishing. This is what the first thread is waiting for, as designed (it will not advance to the next state until this one is complete).

So I guess the question is, why does `sleep()` not return? I will setup a 1-1 topology to stress test such a case. Meanwhile, can you see if switching `sleep(1)` to `std::this_thread::sleep_for(std::chrono::seconds(1))` makes any difference?

**Comment by ktf** [comment]
I suspect that the issue is that the transition is requested via a signal. If you setup a 1-1 topology, could you take that into account?

**Comment by ktf** [comment]
I think there is something seriously wrong in the signal handling code as sometimes FairMQ seems to ignore them as well. Did anything change recently?

**Comment by rbx** [comment]
Last change was 4 months ago.
Many of our tests use signals to stop devices in a similar way that you described - let one device finish and SIGINT the rest from the running script. So far there hasn't been a single fail due to uncaught signals.

So what I will do now is add sleep() to the mix to see if it makes any difference. [Docs on sleep()](https://linux.die.net/man/2/nanosleep) do mention some signals specialties, another reason why I would stay away from it and use std.

**Comment by ktf** [comment]
I will remove for sure all the sleeps, yet I am worried that that's only what triggers some actual issue in the signal handling code.

**Comment by rbx** [comment]
Tested it on Linux - any `sleep()` in the main thread unblocks and returns number of seconds left to sleep. A sleep in an extra thread will ignore the signal and keep sleeping for its specified duration and return 0. No hanging so far. Gonna try MacOS now.

**Comment by rbx** [comment]
For reference, this is what i am testing:
https://github.com/rbx/FairMQ/tree/test-sleep/examples/1-1
Check the test-ex-1-1.sh.in and Sampler.cxx please, is this similar to your scenario?

**Comment by dennisklein** [comment]
Do you record the stacktrace at a point in time, when the sleep should have finished already? Also, do you see this issue on Linux as well?

As to the signal handling: The handler registered with the OS just [sets an atomic global](https://github.com/FairRootGroup/FairMQ/blob/master/fairmq/plugins/Control.cxx#L25-35) and a thread is [observing this atomic by polling it](https://github.com/FairRootGroup/FairMQ/blob/master/fairmq/plugins/Control.cxx#L314). In the example you posted this must have worked, because the condition_variable the first thread hangs on, is in the stop function which was called by our signal handler thread.

I suspect there is some mad interplay between mac sleep and our signal handling, or do you have non sleep examples that are broken? Where can I find docs about the `__semwait_signal` syscall? At least the name suggests it is also observing signals? But even if it was racing with our signal handler, why did it not wake up by clock? Is the clock also sending a signal, that we consume somehow?

**Comment by rbx** [comment]
The test also passes on Mac. But then i guess for you it also breaks only in a fraction of test runs?

**Comment by ktf** [comment]
The issues are also on linux.

**Comment by rbx** [comment]
So far I cannot reproduce this anywhere. I will let `for i in {1..10000000}; do echo -n "Test #$i "; date; examples/1-1/test-ex-1-1.sh; done` run until it fails.

**Comment by dennisklein** [comment]
> I think there is something seriously wrong in the signal handling code as sometimes FairMQ seems to ignore them as well. Did anything change recently?

Can you rule out that your code or one of your dependencies installs their own signal handlers? As far as I know this would explain the symptom of missed signals.

**Comment by ktf** [comment]
Given root is linked in I cannot exclude anything. IIRC, I don’t manipulate signals myself. Notice also that monitoring uses a separate thread as well.

**Comment by rbx** [comment]
We did have some signal handling issue involving root very long time ago (~3 years) in one of FairRoot examples. We ended up disabling root signals handling in that case. Perhaps it is worth trying to further narrow down the problem search:
```C++
        gSystem->ResetSignal(kSigInterrupt);
        gSystem->ResetSignal(kSigTermination);
```

(my test loop still going without fail: `Test #1927 Fr 8. Feb 14:50:42 CET 2019`)

**Comment by rbx** [comment]
Still no fail after 100k iterations...
`Test #113310 Mo 11. Feb 10:28:40 CET 2019`

**Comment by ktf** [comment]
We see the issue something like 1/100, so you are probably not reproducing it correctly. I will try with the gSystem suggestion.

**Comment by ktf** [comment]
Going back to this. I tried the gSystem part and it did not help. Now I have a new suspect: is FairLogger safe to use after a signal is received?

**Comment by ktf** [comment]
To be clear, I suspect that something is already deleted when I invoke LOG() on shutdown, leaving the system in some sort of unstable state...

**Comment by rbx** [comment]
FairLogger 1.4.0 includes a fairly recent [fix](https://github.com/FairRootGroup/FairLogger/pull/11) that fixed a problem when it was used after its static members have been already deleted. I don't think this is directly related to signals, but possibly as a result of the signal you get into a state where the logger is already destructed. With the fix you will still get the output, but unformatted.

Besides this I can't think of anything that a signal would change in the FairLogger.

**Comment by rbx** [comment]
I did some tests with sleep+signals+threads.

[sleep()](url) does some signal handling by itself:
```
sleep() causes the calling thread to sleep either until the number
of real-time seconds specified in seconds have elapsed or
until a signal arrives which is not ignored.
```
I tested this inside FairMQ in a following way: have a `sleep(5)` inside main thread (in `Run()`) and another `sleep(5)` in a separate thread. The `sleep()` from the main thread unblocks with the signal, but the one in a separate keeps sleeping full 5 seconds.

Using the POSIX API there may be a way to unblock any sleep by doing some trickery with the signal handling (by not ignoring them in any thread). But the recommended way is to block signals in all threads except one that is dedicated to signal handling. With this recommendation it sounds to me that the sleeps will keep sleeping for full duration, since signals are blocked (same as "ignored"?).

However we do not use POSIX APIs to handle signals, we use `std::signal()`, where it is not specified which thread will call the handler (although in my tests it is the main thread). Thus I don't see a way to unblock any `sleep()` via the signal handler.

Doing same test with `std::this_thread::sleep_for(chrono::duration)` none of the sleeps are unblocked by the signal and sleep for full specified duration (by design). Switching to `std::this_thread::sleep_for` should at least unblock you (while keeping the behavior you want), since it doesn't do any signal activity.

Of course in your case the sleep neither interrupts nor sleeps full duration, but hangs. My tests still do not reveal what is causing this (nor could i reproduce it).
